### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by vlopes11**

### CODE REVIEW

File: Cargo.toml
Comment (adlerjohn) [inline Cargo.toml:6]:
```suggestion
authors = ["Fuel Labs <contact@fuel.sh>"]
categories = ["cryptography::cryptocurrencies"]
```

File: Cargo.toml
Comment (adlerjohn) [inline Cargo.toml:12]:
```suggestion
description = "A pBFT scheme for Fuel."
```

**Review by adlerjohn — COMMENTED** [review]
Nits. Before merging this PR, can we add a CI that runs tests (if any) and at least checks formatting?

File: Cargo.toml
Comment (vlopes11) [reply Cargo.toml:6]: Why remove the edition?

File: src/block.rs
Comment (Voxelot) [inline src/block.rs:3]: What do people think about making the transaction type a generic trait, rather than directly depending on fuel-tx?

Comment (vlopes11) [reply src/block.rs:3]: That sounds like a good idea

File: src/network.rs
Comment (Voxelot) [inline src/network.rs:29]: I'm wondering if we should even include this. Since the core of pbft operates at the block-level, interactions with the txpool could be handled at the implementation layer of the `Node` trait.

Comment (vlopes11) [reply src/network.rs:29]: We want to avoid serialization/deserialization of types based on agnostic bytes. Using the transaction here is a way to bring this spec closer to our protocol, but it indeed can be replaced by associated types of the trait that satisfies some criteria

Comment (Voxelot) [reply src/network.rs:29]: Since the block type is already generic, and methods like `new_block` and `validate_block` are also generic, I don't see the need for the pbft algorithm to manually handle specific txs. Except in the case of compact blocks, then we'd need to send messages for requesting the missing tx id's from peers. Even then, the pbft algorithm would only need to know if the tx exists and what id it is. Not the actual bytes or payload of it.

Comment (vlopes11) [reply src/network.rs]: This commit that goes into that direction: https://github.com/FuelLabs/fuel-pbft/pull/1/commits/0b5f3bf17a5a33d6342ade339873b0dbee6bee77

File: src/message.rs
Comment (Voxelot) [inline src/message.rs:18]: We should also have `round(&self)` in addition to height.

File: src/transaction.rs
Comment (Voxelot) [inline src/transaction.rs:3]: This type seems overly complicated and doesn't properly restrict the input domain to valid structures. I'm worried about all the edge cases that may arise from handling this recursive structure. Can we just stick with a `Vec<Transaction>` for now?

Comment (vlopes11) [reply src/transaction.rs:3]: That is true. I replaced the transaction entirely by a `Payload` that will impose no restrictions to the contents of the block. https://github.com/FuelLabs/fuel-pbft/blob/76b73e81cc004ddb659ea19dcec47bd67307fcf6/src/block.rs#L8

Comment (vlopes11) [reply src/transaction.rs:3]: We now ask the network to generate a new payload at any given moment, without imposing restrictions:
https://github.com/FuelLabs/fuel-pbft/blob/76b73e81cc004ddb659ea19dcec47bd67307fcf6/src/network.rs#L33-L34

File: Cargo.toml
Comment (SilentCicero) [inline Cargo.toml:8]: "fuel-vm" -> "fuel" ?

File: .github/workflows/ci.yml
Comment (Voxelot) [inline .github/workflows/ci.yml:14]:
```suggestion
  RUSTFLAGS: -D warnings
```

File: .github/workflows/ci.yml
Comment (Voxelot) [inline .github/workflows/ci.yml:44]:
```suggestion

```
We don't really need to build and then test, since test will already build the project.

File: src/node.rs
Comment (Voxelot) [inline src/node.rs:9]: Since peers could be non-validating nodes or light clients, we should go with something more like `AuthorityId` or `ValidatorId`

**Comment by SilentCicero** [comment]
@vlopes11 would be nice to start off with some tests, even basic.

File: tests/mock/src/lib.rs
Comment (SilentCicero) [inline tests/mock/src/lib.rs:3]: Think we could eventually simulate this with `sepk256` eventually? For now this curve is totally fine and even ideal as we mock up everything.

File: src/message.rs
Comment (vlopes11) [reply src/message.rs:12]: It is a bit controversial because the paper won't make references to the consensus implementation of `(H, R)`, but only rounds instead. I think the best course of action is to create a trait `Round` and the implementor can use it as he wants, being that a tuple as in tendermint or a single height.

Thoughts?

File: tests/mock/src/lib.rs
Comment (vlopes11) [reply tests/mock/src/lib.rs:3]: Yes, the choice of curve here is arbitrary

File: src/lib.rs
Comment (Voxelot) [inline src/lib.rs:1]:
```suggestion
#![cfg_attr(not(feature = "std"), no_std)]
```

File: src/message.rs
Comment (Voxelot) [reply src/message.rs]: Our goal is to follow in the footsteps of Tendermint, which is more specifically suited and battle-tested for decentralized blockchain-based protocols. While pBFT is the foundation for the majority of modern approaches and can be cited, I don't think we should restrict our design to a paper from 1999.

The Tendermint paper has a clear definition of height meaning the block height, and rounds which reset to 0 at the start of each new block height.
https://tendermint.com/static/docs/tendermint.pdf

tendermint

File: src/lib.rs
Comment (vlopes11) [reply src/lib.rs:1]: There is no need to use features here since we won't ever use `std` in the crate - only in the implementations

File: src/message.rs
Comment (vlopes11) [reply src/message.rs]: Hmm but if we are to use the tendermint paper as base, then there is another change I need to make. According to tendermint, we propose the new block after the confirmation of `precommit`, whereas we currently do so only after `commit`. cc @adlerjohn 

Its naturally not a challenge to go for the tendermint approach, but then how about we rename the lib?

Comment (Voxelot) [reply src/message.rs]: Renaming the lib might be in order, e.g. fuelmint or something similar as Tendermint is what we had agreed on.

File: src/node.rs
Comment (Voxelot) [inline src/node.rs:3]: We should make these traits async. That way we could use tokio to simulate time or I/O glitches in our unit tests.

Comment (vlopes11) [reply src/node.rs:3]: The core thing of this implementation so far is to be lightweight. If we put the traits async, we will add dependencies and possibly features.

We can alternatively merge this initial commit and then treat these improvements via issues and dedicated PRs. Thoughts?

File: src/lib.rs
Comment (Voxelot) [reply src/lib.rs:1]: why do we want this crate to be no-std?

File: src/node.rs
Comment (Voxelot) [reply src/node.rs:3]: I'm fine with merging without that for now. But eventually, this should include some kind of tokio task or executor that drives progress and handles timeouts etc.

File: src/lib.rs
Comment (vlopes11) [reply src/lib.rs:1]: Its the other way around. There is no reason to make it `std`, so being `no-std` is more powerful and flexible.

File: src/node.rs
Comment (vlopes11) [reply src/node.rs:3]: Yep, we definitely should add an issue to keep track of the requirement

File: src/round.rs
Comment (Voxelot) [inline src/round.rs:8]: What are these for?

File: src/message.rs
Comment (vlopes11) [reply src/message.rs]: To follow-up: https://github.com/FuelLabs/fuel-pbft/issues/2

File: src/round.rs
Comment (vlopes11) [reply src/round.rs:8]: This is a convenience implementation of the round abstraction. If we are under pbft, the rules of tendermint wouldn't specify the ruleset of this crate, and this allows us to implement tendermint natively.

Comment (Voxelot) [reply src/round.rs:8]: I'm just not sure why we need more than (round) or (block height, round)? This genericism may make it unnecessarily complex to implement traits like partialeq etc.

File: src/node.rs
Comment (adlerjohn) [reply src/node.rs:3]: #3

File: src/round.rs
Comment (adlerjohn) [reply src/round.rs:8]: +1 on we don't need things to be too generic. It's just Tendermint.

**Review by Voxelot — APPROVED** [review]
Approving so we can checkpoint this progress.

**Pull request merged**
